File: src/api/private/news-remote-mutate.ts
import { RequestFN } from "~/core/request-function";
import { encodeDomain } from "~/encoders/domain";
import { NewsInformation, type NewsItem, NewsSurvey, type SessionHandle, TabLocation } from "~/models";
import type { NewsQuestion } from "~/models/news-question";
import { NewsQuestionKind } from "~/models/news-question-kind";

/**
 * Updates the status of a news item.
 * Could be a read, or answer to a survey.
 */
export const newsRemoteMutate = async (session: SessionHandle, item: NewsInformation | NewsSurvey, options = {
  delete: false,
  markAsRead: true,
  onlyMarkAsRead: false
}) => {
  const answers = (options.onlyMarkAsRead || options.delete) ? []
    : item.is === "information" ? [item.question] : item.questions;

  const request = new RequestFN(session, "SaisieActualites", {
    _Signature_: { onglet: TabLocation.News },

    donnees: {
      listeActualites: [{
        N: item.id,
        L: item.title ?? "",

        genrePublic: item.public.G,
        public: item.public,

        validationDirecte: true,

        listeQuestions: answers.map((answer) => ({
          N: answer.id,
          L: answer.fullTitle,
          genreReponse: answer.kind,

          reponse: {
            N: parseInt(answer.answerID),
            Actif: true,

            // Should give a string directly when we reply to an information
            // or when the question is just a text input.
            valeurReponse: (answer.kind === NewsQuestionKind.InformationText || answer.kind === NewsQuestionKind.TextInput)
              ? answer.textInputAnswer ?? ""
              : { // Otherwise, pass the choices as an array of numbers.
                _T: 8,
                V: encodeDomain(answer.selectedAnswers ?? [])
              },

            // Defined when the question have choices which can have a text input : "Other" choice.
            valeurReponseLibre: (answer.choices.filter((choice) => choice.isTextInput).length > 0)
              ? answer.textInputAnswer
              : undefined,

            avecReponse: answer.answered,
            estReponseAttendue: answer.shouldAnswer,
            _validationSaisie: true
          }
        })),

        marqueLueSeulement: !options.delete && options.onlyMarkAsRead,
        lue: !options.delete && options.markAsRead,
        supprimee: !options.onlyMarkAsRead && options.delete,
        saisieActualite: false
      }],

      saisieActualite: false
    }
  });

  await request.send();
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/private/news-remote-mutate.kt
package ink.literate.pawnote.api.private

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.encoders.encodeDomain
import ink.literate.pawnote.models.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

data class NewsRemoteMutateOptions(
    val delete: Boolean = false,
    val markAsRead: Boolean = true,
    val onlyMarkAsRead: Boolean = false
)

/**
 * Updates the status of a news item.
 * Could be a read, or answer to a survey.
 */
suspend fun newsRemoteMutate (
    session: SessionHandle,
    item: NewsItem<Any>,
    options: NewsRemoteMutateOptions
) {
    val answers: List<NewsQuestion> = if (options.onlyMarkAsRead || options.delete) listOf() else (if (item.data is NewsInformation) listOf(item.data.question) else (item.data as NewsSurvey).questions)

    val request = RequestFN(session.information, "SaisieActualites", Json.encodeToString(
        buildJsonObject {
            putJsonObject("_Signature_") {
                put("onglet", TabLocation.News.code)
            }

            putJsonObject("donnees") {
                putJsonArray("listeActualites") {
                    addJsonObject {
                        put("N", item.id)
                        put("L", item.title ?: "")

                        put("genrePublic", item.public["G"]!!)
                        put("public", item.public)

                        put("validationDirecte", true)

                        putJsonArray("listeQuestions") {
                            for (answer in answers) {
                                addJsonObject {
                                    put("N", answer.id)
                                    put("L", answer.fullTitle)
                                    put("genreReponse", answer.kind.code)

                                    putJsonObject("reponse") {
                                        put("N", answer.answerId.split('#')[0].toInt())
                                        put("Actif", true)

                                        // Should give a string directly when we reply to an information
                                        // or when the question is just a text input.
                                        if (answer.kind === NewsQuestionKind.InformationText || answer.kind === NewsQuestionKind.TextInput)
                                            put("valeurReponse", answer.textInputAnswer)
                                        else
                                            putJsonObject("valeurReponse") {
                                                put("_T", 8)
                                                put("V", encodeDomain(answer.selectedAnswers ?: listOf()))
                                            }

                                        put("valeurReponseLibre", if (answer.choices.any { it.isTextInput }) answer.textInputAnswer else null)

                                        put("avecReponse", answer.answered)
                                        put("estReponseAttendue", answer.shouldAnswer)
                                        put("_validationSaisie", true)
                                    }
                                }
                            }
                        }

                        put("marqueLueSeulement", !options.delete && options.onlyMarkAsRead)
                        put("lue", !options.delete && options.markAsRead)
                        put("supprimee", !options.onlyMarkAsRead && options.delete)
                        put("saisieActualite", false)
                    }
                }

                put("saisieActualite", false)
            }
        }
    ))

    request.send()
}